File: api/posthog_client.py
import posthog
import os
import json
import requests
from typing import Dict, Any, Optional
from dotenv import load_dotenv

load_dotenv('.env.local')


class PrivatePostHogClient:
    """Client for PostHog private API (server-side events)"""
    
    def __init__(self):
        self.api_key = os.getenv('PRIVATE_POSTHOG_KEY')
        self.host = os.getenv('PRIVATE_POSTHOG_DOMAIN', 'https://us.i.posthog.com')
        
        if self.api_key:
            self.enabled = True
        else:
            print("Warning: PRIVATE_POSTHOG_KEY not set. PostHog private events will not be sent.")
            self.enabled = False
    
    def capture(self, event: str, distinct_id: str, properties: Optional[Dict[str, Any]] = None, timestamp: Optional[str] = None):
        """
        Capture a PostHog event using the private API.
        
        Args:
            event: The event name
            distinct_id: The user ID
            properties: Optional event properties
            timestamp: Optional timestamp in ISO 8601 format
        """
        if not self.enabled:
            print(f"PostHog private disabled - would send event: {event} for user: {distinct_id}")
            if properties:
                print(f"Properties: {properties}")
            return
        
        try:
            url = f"{self.host}/i/v0/e/"
            headers = {
                "Content-Type": "application/json"
            }
            
            payload = {
                "api_key": self.api_key,
                "event": event,
                "distinct_id": distinct_id,
                "properties": properties or {}
            }
            
            if timestamp:
                payload["timestamp"] = timestamp
            
            response = requests.post(url, headers=headers, data=json.dumps(payload))
            response.raise_for_status()
            print(f"PostHog private event sent: {event} for user: {distinct_id}")
        except Exception as e:
            print(f"Error sending PostHog private event: {e}")


# Create global instance
posthog_private_client = PrivatePostHogClient()


def capture_private_event(event: str, distinct_id: str, properties: Optional[Dict[str, Any]] = None, timestamp: Optional[str] = None):
    """
    Convenience function to capture a PostHog event using the private client.
    
    Args:
        event: The event name
        distinct_id: The user ID
        properties: Optional event properties
        timestamp: Optional timestamp in ISO 8601 format
    """
    posthog_private_client.capture(event, distinct_id, properties, timestamp)


def fetch_events(event_name: str = None, after: str = None, before: str = None, limit: int = 100):
    """
    Fetch events from PostHog using the private API.
    Args:
        event_name: Filter by event name (optional)
        after: ISO date string for filtering events after this date (optional)
        before: ISO date string for filtering events before this date (optional)
        limit: Max number of events to fetch (default 100)
    Returns:
        JSON response from PostHog API
    """
    personal_api_key = os.getenv('POSTHOG_PERSONAL_API_KEY')
    project_id = os.getenv('POSTHOG_PROJECT_ID')
    host = os.getenv('PRIVATE_POSTHOG_DOMAIN', 'https://us.posthog.com')
    if not personal_api_key or not project_id:
        raise Exception('POSTHOG_PERSONAL_API_KEY and POSTHOG_PROJECT_ID must be set in environment')
    url = f"{host}/api/projects/{project_id}/events/"
    headers = {
        "Authorization": f"Bearer {personal_api_key}",
        "Content-Type": "application/json"
    }
    params = {
        "limit": limit,
        "format": "json"
    }
    if event_name:
        params["event"] = event_name
    if after:
        params["after"] = after
    if before:
        params["before"] = before
    response = requests.get(url, headers=headers, params=params)
    response.raise_for_status()
    return response.json()

File: api/posthog-client.ts
import axios from 'axios';
import * as dotenv from 'dotenv';
import { PostHogEvent } from './datatypes';

dotenv.config({ path: '.env.local' });

class PrivatePostHogClient {
  private apiKey: string | undefined;
  private host: string;
  public enabled: boolean;

  constructor() {
    this.apiKey = process.env.PRIVATE_POSTHOG_KEY;
    this.host = process.env.PRIVATE_POSTHOG_DOMAIN || 'https://us.i.posthog.com';
    
    if (this.apiKey) {
      this.enabled = true;
    } else {
      console.log("Warning: PRIVATE_POSTHOG_KEY not set. PostHog private events will not be sent.");
      this.enabled = false;
    }
  }

  async capture(
    event: string, 
    distinctId: string, 
    properties?: Record<string, any>, 
    timestamp?: string
  ): Promise<void> {
    /**
     * Capture a PostHog event using the private API.
     * 
     * Args:
     *   event: The event name
     *   distinctId: The user ID
     *   properties: Optional event properties
     *   timestamp: Optional timestamp in ISO 8601 format
     */
    if (!this.enabled) {
      console.log(`PostHog private disabled - would send event: ${event} for user: ${distinctId}`);
      if (properties) {
        console.log(`Properties: ${JSON.stringify(properties)}`);
      }
      return;
    }

    try {
      const url = `${this.host}/i/v0/e/`;
      const headers = {
        "Content-Type": "application/json"
      };

      const payload: any = {
        api_key: this.apiKey,
        event: event,
        distinct_id: distinctId,
        properties: properties || {}
      };

      if (timestamp) {
        payload.timestamp = timestamp;
      }

      const response = await axios.post(url, payload, { headers });
      console.log(`PostHog private event sent: ${event} for user: ${distinctId}`);
    } catch (error) {
      console.error(`Error sending PostHog private event: ${error}`);
    }
  }
}

// Create global instance
const posthogPrivateClient = new PrivatePostHogClient();

export async function capturePrivateEvent(
  event: string, 
  distinctId: string, 
  properties?: Record<string, any>, 
  timestamp?: string
): Promise<void> {
  /**
   * Capture a PostHog event using the private API.
   * This is a convenience function that uses the global posthogPrivateClient instance.
   */
  await posthogPrivateClient.capture(event, distinctId, properties, timestamp);
}

export async function fetchEvents(
  eventName?: string, 
  after?: string, 
  before?: string, 
  limit: number = 100
): Promise<any> {
  /**
   * Fetch events from PostHog using the private API.
   * Args:
   *   eventName: Filter by event name (optional)
   *   after: ISO date string for filtering events after this date (optional)
   *   before: ISO date string for filtering events before this date (optional)
   *   limit: Max number of events to fetch (default 100)
   * Returns:
   *   JSON response from PostHog API
   */
  const personalApiKey = process.env.POSTHOG_PERSONAL_API_KEY;
  const projectId = process.env.POSTHOG_PROJECT_ID;
  const host = process.env.PRIVATE_POSTHOG_DOMAIN || 'https://us.posthog.com';

  console.log(`Environment check - personalApiKey: ${personalApiKey ? 'set' : 'NOT SET'}`);
  console.log(`Environment check - projectId: ${projectId ? 'set' : 'NOT SET'}`);
  console.log(`Environment check - host: ${host}`);

  if (!personalApiKey) {
    throw new Error('POSTHOG_PERSONAL_API_KEY environment variable is not set');
  }
  if (!projectId) {
    throw new Error('POSTHOG_PROJECT_ID environment variable is not set');
  }

  const url = `${host}/api/projects/${projectId}/events/`;
  const headers = {
    "Authorization": `Bearer ${personalApiKey}`,
    "Content-Type": "application/json"
  };
  
  const params: Record<string, any> = {
    limit: limit,
    format: "json"
  };
  
  if (eventName) {
    params.event = eventName;
  }
  if (after) {
    params.after = after;
  }
  if (before) {
    params.before = before;
  }

  console.log(`Making request to: ${url}`);
  console.log(`With params: ${JSON.stringify(params)}`);

  try {
    const response = await axios.get(url, { headers, params });
    console.log(`Response status: ${response.status}`);

    if (response.status !== 200) {
      console.log(`Response text: ${response.data}`);
      throw new Error(`HTTP ${response.status}: ${response.statusText}`);
    }

    return response.data;
  } catch (error) {
    console.error(`Error fetching events: ${error}`);
    throw error;
  }
} 